enhancements/olm-deletes-operands.md

```diff
@@ -0,0 +1,294 @@
+
+#### Story 1
+
+As an operator user I want my operator uninstall to also clean up operator-created resources so that I don't have to manually clean up any leftover resources.
+
+### Implementation Details/Notes/Constraints
+
+#### Where to configure cleanup: CSV and/or Subscription
+
+The API to configure operand cleanup could be specified on either the Subscription or ClusterServiceVersion (CSV).
+
+If the cleanup API is only present on the Subscription then operators installed without Subscriptions (by manually creating the CSV and fulfilling requirements) can not opt-in to operand cleanup.
+
+The CSV is the more appropriate point to configure the cleanup option for an operator as the CSV represents the actual installation of the operator, whereas the Subscription only represents the intent to subscribe to updates from the catalog. Moreover, since deleting the CSV causes the operator to be removed, deleting the CSV should be the trigger to initiate operand cleanup. See the [section on CSV finalizers][finalizers-on-csvs] for more details.
+
+#### Deleting CRs
+
+The operand cleanup process, simply put, is the removal of all CRs being reconciled by an operator so it can run finalizers for each CR and clean up any associated operator-created resources.
+
+From OLM's perspective these would be all CRs in the `OperatorGroup`'s [target namespaces][target-namespaces] that are of the same type as the [owned CRDs][owned-crds] listed in the CSV. This represents the set of all CRs that the operator reconciles and is responsible for cleaning up after.
+
+OLM should not delete CRs of the type [required CRDs][required-crds] as those are likely created by the operator, and whose cleanup should be handled by the operator's finalizer.
+
+The CRs must directly be deleted instead of only deleting the CRD and relying on garbage collection of CRs because of the complication that causes for namespace scoped operators. See the [CRD removal section][crd-removal-section] below for more details.  
+
+#### Finalizers on CSVs
+
+OLM can initiate the process of operand clean up when the operator is uninstalled, which is when the CSV is deleted. However, deleting a CSV causes the operator deployment (owned by the CSV) to be garbage collected.
+
+So in order to keep the operator around to process CR deletion events, OLM can add a finalizer to the CSV to prevent the CSV and the operator from being removed. Once OLM has completed operand cleanup and can verify the removal of all CRs, the CSV finalizer can be cleared so that the operator can be safely removed.
+
+```YAML
+apiVersion: operators.coreos.com/v1alpha1
+kind: ClusterServiceVersion
+...
+metadata:
+  ...
+  finalizers:
+  - operatorframework.io/cleanup-apis
+```
+
+In the event of operand cleanup being stalled due to some operator error, OLM will not clear the CSV finalizer so that the operator stays alive indefinitely. This way the operator can be observed to figure out the required action to unblock cleanup.
```

Comment:
Can the user unblock the uninstall by going back and editing the CSV to opt-out of operand cleanup?

Reply:
Yeah I forgot to explicitly mention that but we could do that. 
However this would be more like abort cleanup to unblock operator uninstall. No guarantees on whether OLM has already deleted all CRs or not.

Reply:
yeah, that's a fair characterization.  Just want to be sure that:
1) we have a way to finish the uninstall
2) we describe the mechanism (maybe add a user story for it?)

Reply:
I'll add a user story and outline that case in more detail.